src/D2L.CodeStyle.Analyzers/Language/LikelyArgumentMismatchAnalyzer.cs

```diff
@@ -0,0 +1,773 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class LikelyArgumentMismatchAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.LikelyArgumentMismatch
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalyzer );
+		}
+
+		private static void RegisterAnalyzer( CompilationStartAnalysisContext context ) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeInvocation(
+					ctx,
+					(InvocationExpressionSyntax)ctx.Node
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		private static void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			InvocationExpressionSyntax expr
+		) {
+			if( expr.ArgumentList == null ) {
+				return;
+			}
+
+			if( expr.ArgumentList.Arguments.Count <= 1 ) {
+				return;
+			}
+
+			// get bipartite matching of the source
+			ImmutableArray<Edge> sourceMatching = MatchArguments(
+				context.SemanticModel,
+				expr.ArgumentList.Arguments
+			).ToImmutableArray();
+
+			if( sourceMatching.Length <= 1 ) {
+				return;
+			}
+
+			ImmutableArray<Edge> bestMatching = GenerateBestMatching( context, expr );
+
+			// We weren't able to find a better matching
+			if( bestMatching == null ) {
+				return;
+			}
+
+			ImmutableDictionary<ArgumentDetails, Edge> sourceEdges = sourceMatching
+				.ToImmutableDictionary( x => x.Arg, x => x );
+			foreach( var edge in bestMatching ) {
+				Edge sourceEdge = sourceEdges[ edge.Arg ];
+				if( edge.Param != sourceEdge.Param ) {
+					context.ReportDiagnostic( Diagnostic.Create(

```

Comment:
Will this generate two warnings? Should we have a `break` once if finds one diagnostic?